vmm: Allow simultaneously set serial and console as TTY mode

zhuangel opened the pull request:
Performance of legacy serial is poor, so we want to use virtio console for better performance, but once we configure VM console as virito console, the boot stage kernel logs will be lost, including some critical exception message in boot stage.

Linux support assign console for early stage via "earlyprintk", with this feature we could assign legacy serial to early stage, and virito console to normal console (after early boot stage), then we could catch most VM console logs via virito console, also catch boot stage important logs via legacy serial.

This patch remove the double TTY restriction, so that we could use both serial and console with TTY mode.

alyssais commented on vmm/src/config.rs:
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
         if self.console.mode == ConsoleOutputMode::Tty && self.serial.mode == ConsoleOutputMode::Tty
         {
-            return Err(ValidationError::DoubleTtyMode);
+            info!("Notice, {}", ValidationError::DoubleTtyMode);
```
```suggestion
            info!("{}", ValidationError::DoubleTtyMode);
```

No other info message in Cloud Hypervisor is prefixed with "Notice".

zhuangel replied:
Thanks will update.

likebreath reviewed (commented):
Can you please add an integration tests for this particular use-case?

Typo in the commit message: "enable both legacy serial `an` virito console"

zhuangel commented:
> Can you please add an integration tests for this particular use-case?
> 
> Typo in the commit message: "enable both legacy serial `an` virito console"

Thanks for review, will update.

rbradford commented:
Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...

zhuangel commented:
> Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...

Thanks, I have updated the commit message.

rbradford commented:
> > Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...
> 
> Thanks, I have updated the commit message.

This looks a lot better thank you - but I think you meant "mode" rather than "model" in the summary.

zhuangel commented:
> > > Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...
> > 
> > 
> > Thanks, I have updated the commit message.
> 
> This looks a lot better thank you - but I think you meant "mode" rather than "model" in the summary.

Thanks, it should be "mode", updated :)

likebreath commented on tests/integration.rs:
```diff
@@ -7093,6 +7093,58 @@ mod common_parallel {
+        let guest = Guest::new(Box::new(focal));
+        let mut cmd = GuestCommand::new(&guest);
+        let api_socket = temp_api_path(&guest.tmp_dir);
+        let tty_str: &str = "console=hvc0 earlyprintk=ttyS0 ";
```
@michael2012z Are you aware of anything equivalent to `earlyprintk` on `aarch64`?

likebreath commented on vmm/src/config.rs:
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
             }
         }
 
+        // The 'conflict' check is introduced in commit 24438e0390d3
+        // (vm-virtio: Enable the vmm support for virtio-console).
+        //
+        // Allow double tty mode, for someone want to use virtio console
+        // for better performance, and want to keep legacy serial to catch
+        // boot stage logs for debug.
+        // Using such double tty mode, you need to configure the kernel
+        // properly, such as:
+        // "console=hvc0 earlyprintk=ttyS0"
```
The comments also need to be updated to avoid confusions on "double tty mode".

zhuangel replied:
Thanks, updated.

likebreath commented on vmm/src/config.rs:
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
         if self.console.mode == ConsoleOutputMode::Tty && self.serial.mode == ConsoleOutputMode::Tty
         {
-            return Err(ValidationError::DoubleTtyMode);
+            info!("{}", ValidationError::DoubleTtyMode);
```
I think it is better to say something like: `Both serial and console are being set as TTY mode. Guest kernel command-line should include "console=hvc0 earlyprintk=ttyS0".`

zhuangel replied:
I don't think we should add linux kernel command-line suggestion in CH logs（maybe in doc or comments）, so just change the error return into some logs is enough.

zhuangel commented on vmm/src/config.rs:
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
         {
-            return Err(ValidationError::DoubleTtyMode);
+            info!("{}", ValidationError::DoubleTtyMode);
         }
```
Thanks, updated.

rbradford reviewed (approved):
I'm happy with this as is - @likebreath ?
